### PR TITLE
FOUNDATIONS: Modify User

### PR DESCRIPTION
**Pull request by iamDiyorjon**

### CODE REVIEW

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 1]
```diff
@@ -0,0 +1,278 @@
+﻿//=================================
+// Copyright (c) Coalition of Good-Hearted Engineers
+// Free to use to bring order in your workplace
+//=================================
+
+using EFxceptions.Models.Exceptions;
+using Microsoft.Data.SqlClient;
+using Microsoft.EntityFrameworkCore;
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:8]: order

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+            var failedUserStorageException =
+                new FailedUserStorageException(sqlException);
+
+            var expectedUserDependencyException =
+                new UserDependencyException(failedUserStorageException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+               broker.GetCurrentDateTime())
+                   .Throws(sqlException);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:38]: 2 line and before broker need tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+            //when
+            ValueTask<User> modifyUserTask =
+                this.userService.ModifyUserAsync(randomUser);
+
+            //then
+            await Assert.ThrowsAsync<UserDependencyException>(() =>
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:45]: upgrade new standard

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            var expectedUserDependencyValidationException =
+                new UserDependencyValidationException(lockedUserException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.SelectUserByIdAsync(userId))
+                    .ThrowsAsync(databaseUpdateConcurrencyException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime())
+                    .Returns(randomDateTime);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:197]: 2 line

File: Tarteeb.Api/Services/Foundations/Users/UserService.Validations.cs [hunk 5]
```diff
@@ -37,14 +40,55 @@ private void ValidateUser(User user)
...
+
+        private static void ValidateStorageUser(User maybeUser, Guid userId)
         {
             if (maybeUser is null)
             {
                 throw new NotFoundUserException(userId);
             }
         }
 
+        private void ValidationUserOnModify(User user)
```
Comment (jamshidbek001) [inline Tarteeb.Api/Services/Foundations/Users/UserService.Validations.cs:69]: ValidateUserOnModify bo'lishi kerak

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+            SqlException sqlException = GetSqlException();
+
+            var failedUserStorageException =
+                new FailedUserStorageException(sqlException);
+
+            var expectedUserDependencyException =
+                new UserDependencyException(failedUserStorageException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+               broker.GetCurrentDateTime()).Throws(sqlException);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:35]: Tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+            this.dateTimeBrokerMock.Verify(broker =>
+               broker.GetCurrentDateTime(), Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogCritical(It.Is(SameExceptionAs(
+                    expectedUserDependencyException))), Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+               broker.SelectUserByIdAsync(randomUser.Id), Times.Never);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:55]: tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTime(), Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedUserDependencyValidationException))), Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+               broker.SelectUserByIdAsync(foreignKeyConflictedUser.Id),
+                   Times.Never);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:105]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime()).Returns(randomDateTime);
+
+            // when
+            ValueTask<User> modifyUserTask =
+                this.userService.ModifyUserAsync(someUser);
+
+            UserDependencyException actualUserDependencyException =
+              await Assert.ThrowsAsync<UserDependencyException>(modifyUserTask.AsTask);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.Modify.cs:145]: tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.Add.cs [hunk 10]
```diff
@@ -19,7 +19,7 @@ public partial class UserServiceTests
         public async Task ShouldAddUserAsync()
         {
             //given
-            DateTimeOffset randomDateTime = GetRandomDateTime();
+            DateTimeOffset randomDateTime = GetRandomDateTimeOffset();
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.Add.cs:22]: It should be GetRandomDateTime without Offset

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.Modify.cs [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                await this.userService.ModifyUserAsync(inputUser);
+
+            //then
+            actualUser.Should().BeEquivalentTo(exceptedUser);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTime(), Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectUserByIdAsync(userId),
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.Modify.cs:53]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.Modify.cs [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                    Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.UpdateUserAsync(inputUser),
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.Modify.cs:57]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Validation.Modify.cs [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdateDateIsSameAsCreateDateAndLogItAsync()
+        {
+            //given
+            DateTimeOffset randomDatetime = GetRandomDateTimeOffset();
+            User randomUser = CreateRandomUser(randomDatetime);
+            User invalidUser = randomUser;
+
+            var invalidUserException =
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Validation.Modify.cs:127]: 1 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Validation.Modify.cs [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+
+            invalidUserException.AddData(
+                key: nameof(User.UpdatedDate),
+                values: $"Date is same as {nameof(User.CreatedDate)}");
+
+            var expectedUserValidationException =
+                new UserValidationException(invalidUserException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+               broker.GetCurrentDateTime()).Returns(randomDatetime);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Validation.Modify.cs:138]: tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Validation.Modify.cs [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        [MemberData(nameof(MinutsBeforeOrAfter))]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsNotRecentAndLogItAsync(int minuts)
+        {
+            //given
+            DateTimeOffset dateTime = GetRandomDateTimeOffset();
+            User randomUser = CreateRandomUser(dateTime);
+            User inputUser = randomUser;
+            inputUser.UpdatedDate = dateTime.AddMinutes(minuts);
+
+            var invalidUserException =
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Validation.Modify.cs:175]: 1 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.cs [hunk 16]
```diff
@@ -76,17 +95,22 @@ private static IQueryable<User> CreateRandomUsers()
...
-        private static User CreateRandomUser(DateTimeOffset dates) =>
-            CreateUserFiller(dates).Create();
-
-        private static DateTimeOffset GetRandomDateTimeOffset() =>
-            new DateTimeRange(earliestDate: DateTime.UnixEpoch).GetValue();
+        public static TheoryData MinutsBeforeOrAfter()
+        {
+            int randomNumber = GetRandomNumber();
+            int randomNegativeNumber=GetRandomNegativeNumber();
+            return new TheoryData<int>
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.cs:108]: need empty line

**Review by jamshidbek001 — CHANGES_REQUESTED** [review]
please correct tabs

**Review by StudentofTATU — APPROVED** [review]
LGTM

**Review by jamshidbek001 — APPROVED** [review]

**Pull request merged**